server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java

```diff
@@ -174,53 +174,51 @@ private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, Rout
                 );
             }
 
-            int currentNodeCount = getCurrentNodeCountForAttribute(shardRouting, node, allocation, moveToNode, awarenessAttribute);
+            int numShardCopiesAllocatedToCurrentNodeAttributeValue = getNumberOfShardCopiesAllocatedToCurrentNodeAttributeValue(shardRouting, node, allocation, moveToNode, awarenessAttribute);
 
-            // build attr_value -> nodes map
-            Set<String> nodesPerAttribute = allocation.routingNodes().nodesPerAttributesCounts(awarenessAttribute);
-            int numberOfAttributes = nodesPerAttribute.size();
+            Set<String> attributeValues = allocation.routingNodes().nodesPerAttributesCounts(awarenessAttribute);
+            int distinctAttributeValuesCount = attributeValues.size();
             List<String> fullValues = forcedAwarenessAttributes.get(awarenessAttribute);
 
             if (fullValues != null) {
-                // If forced awareness is enabled, numberOfAttributes = count(distinct((union(discovered_attributes, forced_attributes)))
+                // If forced awareness is enabled, distinctAttributeValuesCount = count(distinct((union(discovered_attributes, forced_attributes)))
                 Set<String> attributesSet = new HashSet<>(fullValues);
-                for (String stringObjectCursor : nodesPerAttribute) {
-                    attributesSet.add(stringObjectCursor);
-                }
-                numberOfAttributes = attributesSet.size();
+                attributesSet.addAll(attributeValues);
```

Comment:
Can we check once how addAll performs vs a single add call on a Set within a loop? Since AllocationDeciders are called on allocating each shard on each node.